pageserver/src/walingest.rs

```diff
@@ -1355,8 +1355,10 @@ impl WalIngest {
             // Note: The multixact members can wrap around, even within one WAL record.
             offset = offset.wrapping_add(n_this_page as u32);
         }
-        if xlrec.mid >= self.checkpoint.nextMulti {
-            self.checkpoint.nextMulti = xlrec.mid + 1;
+        if xlrec.mid.wrapping_sub(self.checkpoint.nextMulti) as i32 >= 0 {
+            let next_mid =
+                std::cmp::max(xlrec.mid.wrapping_add(1), pg_constants::FIRST_MULTIXACT_ID);
+            self.checkpoint.nextMulti = next_mid;
             self.checkpoint_modified = true;
         }
```

Comment:
Let's add a comment here:
```
  // Advance `nextMulti` in wraparound-aware way. This should match the MultiXactAdvanceNextMXact()
  // logic in PostgreSQL's xlog_redo() function.
```

Reply:
The corresponding Postgres logic in xlog_redo doesn't do a comparison with FirstMultixactId, why do we need it here?

(Hmm, shouldn't Postgres need this too, actually?)

Reply:
> Let's add a comment here:
Done

> The corresponding Postgres logic in xlog_redo doesn't do a comparison with FirstMultixactId, why do we need it here?

Sorry, do not completely understand which logic do you mean.
`FirstMultixactId` is taken into account in `SetMultiXactIdLimit`:

```
	/*
	 * We pretend that a wrap will happen halfway through the multixact ID
	 * space, but that's not really true, because multixacts wrap differently
	 * from transaction IDs.  Note that, separately from any concern about
	 * multixact IDs wrapping, we must ensure that multixact members do not
	 * wrap.  Limits for that are set in SetOffsetVacuumLimit, not here.
	 */
	multiWrapLimit = oldest_datminmxid + (MaxMultiXactId >> 1);
	if (multiWrapLimit < FirstMultiXactId)
		multiWrapLimit += FirstMultiXactId;
```

Reply:
Ah, gotcha, I missed that in SetMultiXactIdLimit

Reply:
Oh but wait, the WAL redo routine of XLOG_MULTIXACT_CREATE_ID doesn't call SetMultiXactIdLimit.

Reply:
As far as I understand `nextMXact`is not expected to point to there valid ID.
```
        /*
	 * We don't care about MultiXactId wraparound here; it will be handled by
	 * the next iteration.  But note that nextMXact may be InvalidMultiXactId
	 * or the first value on a segment-beginning page after this routine
	 * exits, so anyone else looking at the variable must be prepared to deal
	 * with either case.  Similarly, nextOffset may be zero, but we won't use
	 * that as the actual start offset of the next multixact.
	 */
	(MultiXactState->nextMXact)++;
```

Everywhere where it is used, it is compared with `FirstMultiXactId`:
```
   	        /*
		 * We have to beware of the possibility that nextMXact is in the
		 * wrapped-around state.  We don't fix the counter itself here, but we
		 * must be sure to store a valid value in our array entry.
		 */
		nextMXact = MultiXactState->nextMXact;
		if (nextMXact < FirstMultiXactId)
			nextMXact = FirstMultiXactId;
```

IMHO seems to be strange decision, because it is advanced in one place and used oil many places, but at  least there seems to be no error in Postgres.

Reply:
Huh, yeah, that's weird. Indeed, Postgres can have nextMulti==0 in a checkpoint, I confirmed that:

```
postgres=# update tt set i = i; checkpoint; select next_multixact_id, next_multi_offset, next_xid from pg_control_checkpoint ();
UPDATE 1
CHECKPOINT
 next_multixact_id | next_multi_offset | next_xid
-------------------+-------------------+----------
        4294967294 |        4294967048 | 0:1012
(1 row)
    
postgres=# update tt set i = i; checkpoint; select next_multixact_id, next_multi_offset, next_xid from pg_control_checkpoint ();
UPDATE 1
CHECKPOINT
 next_multixact_id | next_multi_offset | next_xid
-------------------+-------------------+----------
        4294967295 |        4294967050 | 0:1013
(1 row)
    
postgres=# update tt set i = i; checkpoint; select next_multixact_id, next_multi_offset, next_xid from pg_control_checkpoint ();
UPDATE 1
CHECKPOINT
 next_multixact_id | next_multi_offset | next_xid
-------------------+-------------------+----------
                 0 |        4294967052 | 0:1014
(1 row)
    
postgres=# update tt set i = i; checkpoint; select next_multixact_id, next_multi_offset, next_xid from pg_control_checkpoint ();
UPDATE 1
CHECKPOINT
 next_multixact_id | next_multi_offset | next_xid
-------------------+-------------------+----------
                 2 |        4294967054 | 0:1015
(1 row)

```

I don't see any harm in skipping over 0 in pageserver already, but that's not how Postgres does it, so probably better not to. I'd like to change that in Postgres though, I think it'd be better to skip over 0 already when updating nextMulti, not when reading it.